components/Starknet/modules/architecture_and_concepts/pages/Smart_Contracts/universal-deployer.adoc

```diff
@@ -0,0 +1,124 @@
+:deploy-syscall: xref:./system-calls-cairo1.adoc#deploy[deploy syscall]
+
+[id="universal_deployer"]
+= Universal Deployer Contract
+
+The Universal Deployer Contract (UDC) is a singleton smart contract that wraps the {deploy-syscall} to expose it to any contract that doesn’t implement it, such as account contracts. You can think of it as a standardized generic factory for Starknet contracts.
+
+And since Starknet has no deployment transaction type, it offers a canonical way to deploy smart contracts by following the https://community.starknet.io/t/snip-deployer-contract-interface/2772[standard deployer interface] and emitting a `ContractDeployed` event. For more details on thought process and decision making, refer to the https://community.starknet.io/t/universal-deployer-contract-proposal/1864[Universal Deployer Contract Proposal].
+
+NOTE: The UDC address is `0x041a78e741e5af2fec34b695679bc6891742439f7afb8484ecd7766661ad02bf` in mainnet, testnets, and starknet-devnet. This address may change in the future as it gets migrated to Cairo 1+.
+
+== Interface
+
+[,js]
+----
+trait IUniversalDeployer {
+    fn deployContract(
+        class_hash: ClassHash,
+        salt: felt252,
+        unique: bool,
+        calldata: Span<felt252>
+    ) -> ContractAddress;
+}
+----
+
+== Usage
+
+To deploy a contract first you need to make sure it's been xref:Network_Architecture/transactions.adoc#declare-transaction[declared] and then send a `deployContract` transaction to the Universal Deployer Contract.
+Here's an example implementation in Cairo:
+
+[,js]
+----
+use openzeppelin::utils::universal_deployer::IUniversalDeployerDispatcher;
+
+const UDC_ADDRESS = 0x041a78e741e5af2fec34b695679bc6891742439f7afb8484ecd7766661ad02bf;
+let udc = IUniversalDeployerDispatcher { contract_address: UDC_ADDRESS }
+
+let class_hash = 0x5c478ee27f2112411f86f207605b2e2c58cdb647bac0df27f660ef2252359c6;
+let salt = 1234567879;
+let unique = false;
+let mut calldata = ArrayTrait::<felt252>::new();
+
+let deployed_addr = udc.deployContract(class_hash, salt, unique, calldata.span());
+----
+
+=== Deployment types
+
+The Universal Deployer Contract offers two types of addresses to deploy: origin dependent and origin independent.
+As the names suggest, the origin-dependent type includes the deployer's address in the address calculation;
+whereas, the origin-independent type does not.
+The `unique` boolean parameter ultimately determines the type of deployment.
+
+CAUTION: When deploying a contract that uses `get_caller_address` in the constructor calldata, remember that the UDC (and not the account) deploys that contract.
+Therefore, querying `get_caller_address` in a contract's constructor will return the UDC's address and *not* the account's address.
+
+==== Origin dependent
+
+Making deployments dependent of the origin address, users can "reserve" themselves a whole address space from squatting by someone else.
```

Comment:
1. Who is the user here? Isn't this for developers?
2. Why is "reserve" in quotation marks? Is it a special term in this context? Quotation marks signifies sarcasm, so we don't use them except for specific cases that require them. So I'm not sure what you're trying to say.
3. It says "dependent of the origin address". Do you mean "dependent on the origin address" or "independent of the origin address"?

Reply:
1. the user is whoever uses (calls) the contract. any account can directly call this contract, so you could call it from your wallet if you wanted to. it's not constrained to programatic use by developers.
2. we can drop the quotations marks
3. dependent on the origin address

Reply:
OK. PTAL:

```
By making deployments dependent upon the origin address, users can reserve a whole address space to prevent someone else from taking ownership of the address.

Only the owner of the origin address can deploy to those addresses.
```